napari_plugin_manager/base_qt_plugin_dialog.py

```diff
@@ -0,0 +1,1671 @@
+        self.refresh_button.setDisabled(False)
+
+        if not self.isVisible():
+            if sum(exit_codes) > 0:
+                self._show_warning(
+                    self._trans(
+                        'Plugin Manager: process completed with errors\n'
+                    )
+                )
+            else:
+                self._show_info(
+                    self._trans('Plugin Manager: process completed\n')
+                )
+
+    def _add_to_available(self, pkg_name):
+        self._add_items_timer.stop()
+        self._plugin_queue.insert(0, self._plugin_data_map[pkg_name])
+        self._add_items_timer.start()
+        self._update_plugin_count()
+
+    def _add_installed(self, pkg_name=None):
+        """
+        Add plugins that are installed to the dialog.
+
+        This should call the `installed_list.addItem` method to add each plugin item
+        that should be shown as an installed plugin.
+
+        Parameters
+        ----------
+        pkg_name : str, optional
+            The name of the package that needs to be shown as installed.
+            The default is None. Without passing a package name the logic should
+            fetch/get the info of all the installed plugins and add them to the dialog
+            via the `installed_list.addItem` method.
+
+        """
+        # TODO: This could be better reused?
+        raise NotImplementedError
+
+    def _fetch_available_plugins(self, clear_cache: bool = False):
+        """
+        Fetch plugina available for installation.
+
+        This should call `_handle_yield` in order to queue the addition of plugins available
+        for installation to the corresponded list (`available_list`).
+
+        Parameters
+        ----------
+        clear_cache : bool, optional
+            If a cache is implemented, if the cache should be cleared or not.
+            The default is False.
+
+        """
+        raise NotImplementedError
+
+    def _loading_gif(self):
+        """
+        Movie to use to indicate something is loading.
+
+        This should return an instance of `QMovie` with a scaled size fo 18x18
+
+        """
+        raise NotImplementedError
+
+    def _on_bundle(self):
+        """
+        If the current installation comes from a bundle/standalone approach or not.
+
+        This should return a `bool`, `True` if under a bundle like installation, `False`
+        otherwise.
+        """
+        raise NotImplementedError
+
+    def _show_info(self, info):
+        """
+        Shows a info message.
+
+        Parameters
+        ----------
+        info : str
+            Info message to be shown.
+
+        """
+        raise NotImplementedError
+
+    def _show_warning(self, warning):
+        """
+        Shows a warning message.
+
+        Parameters
+        ----------
+        warning : str
+            Warning message to be shown.
+
+        """
+        raise NotImplementedError
+
+    def _trans(self, text, **kwargs):
```

Comment:
Does it return `str`?

Reply:
Yep, or at least a subclass of `str` so I guess this could be set to describe that it returns something like `TypeVar('T', bound=str)`? I think that should correspond with what napari `trans._` returns (`Union[TranslationString, str]`), right? 🤔